### PR TITLE
Node-ify Temperature

### LINKED ISSUE #146: Configuration temperature control (new Module)
Allow control over Configuration temperature as a function of iteration, providing schedules for heating / cooling.

New module - TemperatureControlModule?

### PR DESCRIPTION
**Pull request by jws-1**
This PR moves the Temperature for configurations from being a GUI control to a node: `Temperature`. For now, we still read `TemperatureKeywords`, but with a deprecation warning. Closes #146.

### CODE REVIEW

File: src/main/keywords_configuration.cpp [hunk 1]
```diff
@@ -77,6 +77,7 @@ bool ConfigurationBlock::parse(LineParser &parser, Dissolve *dissolve, Configura
                 cfg->setRequestedSizeFactor(parser.argd(1));
                 break;
             case (ConfigurationBlock::TemperatureKeyword):
+                Messenger::warn("The TemperatureKeyword will be deprecated in future versions.\n");
```
Comment (trisyoungs) [inline src/main/keywords_configuration.cpp:80]:
```suggestion
                Messenger::warn("The 'Temperature' keyword will be deprecated in a future version.\n");
```

File: src/procedure/nodes/temperature.cpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+// Copyright (c) 2023 Team Dissolve and contributors
+
+#include "procedure/nodes/temperature.h"
+#include "classes/configuration.h"
+#include "keywords/nodeValue.h"
+#include "procedure/nodes/node.h"
+TemperatureProcedureNode::TemperatureProcedureNode()
+    : ProcedureNode(ProcedureNode::NodeType::Temperature, {ProcedureNode::GenerationContext})
+{
+    keywords_.add<NodeValueKeyword>("Temperature", "Temperature in K", temperature_, this);
```
Comment (trisyoungs) [inline src/procedure/nodes/temperature.cpp:11]:
```suggestion
    keywords_.add<NodeValueKeyword>("Temperature", "Temperature (K)", temperature_, this);
```

File: src/procedure/nodes/temperature.cpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+/*
+ * Identity
+ */
+bool TemperatureProcedureNode::mustBeNamed() const { return false; }
```
Comment (trisyoungs) [inline src/procedure/nodes/temperature.cpp:17]:
```suggestion

bool TemperatureProcedureNode::mustBeNamed() const { return false; }
```

File: src/procedure/nodes/temperature.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+/*
+ * Execute
+ */
+// Execute node
```
Comment (trisyoungs) [inline src/procedure/nodes/temperature.cpp:22]:
```suggestion

// Execute node
```

**Review by trisyoungs — CHANGES_REQUESTED** [review]
Looks good, but I think there's a bit more useful work we should do so as not to unnecessarily confuse our users!

First, from the main menu when a user creates a new `Configuration` we need to add our new `TemperatureNode` at the top of the procedure - see `gui/addConfigurationDialogFuncs.cpp`. It should go in at around line 183.

Second, I think we should add a new menu item to allow the temperature of an existing configuration to be set. We shouldn't force the user to regenerate the config from the beginning if they need to tweak the temperature!

File: src/gui/addConfigurationDialogFuncs.cpp [hunk 5]
```diff
@@ -179,6 +180,10 @@ void AddConfigurationDialog::finalise()
     // Create the Configuration and a suitable generator
     auto *newConfiguration = dissolve_.addConfiguration();
     auto &generator = newConfiguration->generator();
+
+    // Add Temperature node, by default
+    generator.createRootNode<TemperatureProcedureNode>("Temperature");
```
Comment (trisyoungs) [inline src/gui/addConfigurationDialogFuncs.cpp:185]: I'd just default this one so it gets the standard node naming
```suggestion
    generator.createRootNode<TemperatureProcedureNode>({});
```

File: src/gui/gui.ui [hunk 6]
```diff
@@ -802,6 +804,11 @@
     <string>Create...</string>
    </property>
   </action>
+  <action name="ConfigurationAdjustTemperatureAction">
+   <property name="text">
+    <string>Adjust Temperature...</string>
```
Comment (trisyoungs) [inline src/gui/gui.ui:809]:
```suggestion
    <string>Adjust &amp;Temperature...</string>
```

File: src/gui/menu_configuration.cpp [hunk 7]
```diff
@@ -136,3 +137,24 @@ void DissolveWindow::on_ConfigurationExportToXYZAction_triggered(bool checked)
...
+{
+    // Get the currently-displayed Configuration
+    auto *cfg = ui_.MainTabs->currentConfiguration();
+    if (!cfg)
+        return;
+
+    auto ok = false;
+    auto temperature = QInputDialog::getDouble(this, "Set temperature for configuration",
+                                               "Enter the temperature (K) to apply to this configuration", cfg->temperature(),
+                                               -10000.0, 10000.0, 3, &ok);
```
Comment (trisyoungs) [inline src/gui/menu_configuration.cpp:151]: Well, that's breaking the laws of physics!  :smiley:
```suggestion
                                               0.0, 1000000.0, 3, &ok);
```

File: src/procedure/nodes/temperature.cpp [hunk 8]
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2023 Team Dissolve and contributors
+
+#include "procedure/nodes/temperature.h"
+#include "classes/configuration.h"
+#include "keywords/nodeValue.h"
+#include "procedure/nodes/node.h"
+TemperatureProcedureNode::TemperatureProcedureNode()
```
Comment (trisyoungs) [inline src/procedure/nodes/temperature.cpp:8]: All hail the BLP!
```suggestion

TemperatureProcedureNode::TemperatureProcedureNode()
```

File: src/procedure/nodes/temperature.cpp [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    : ProcedureNode(ProcedureNode::NodeType::Temperature, {ProcedureNode::GenerationContext})
+{
+    keywords_.add<NodeValueKeyword>("Temperature", "Temperature (K)", temperature_, this);
+}
+
+/*
+ * Identity
+ */
+
+bool TemperatureProcedureNode::mustBeNamed() const { return false; }
```
Comment (trisyoungs) [inline src/procedure/nodes/temperature.cpp:18]: Convention.
```suggestion
// Return whether a name for the node must be provided
bool TemperatureProcedureNode::mustBeNamed() const { return false; }
```

File: src/procedure/nodes/temperature.h [hunk 10]
```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2023 Team Dissolve and contributors
+
+#pragma once
+
+#include "keywords/nodeValue.h"
+#include "procedure/nodes/node.h"
+class TemperatureProcedureNode : public ProcedureNode
```
Comment (trisyoungs) [inline src/procedure/nodes/temperature.h:8]:
```suggestion

class TemperatureProcedureNode : public ProcedureNode
```

**Review by trisyoungs — APPROVED** [review]
Again, just minor stuff. :+1:

We now don't display the temperature on the `ConfigurationTab`, but as long as we're anticipating the tab redesign to be done in the near future no problem.  If not, then adding a label in during the interim is the way to go.

File: src/gui/menu_configuration.cpp (hunk 7 above)
Comment (jws-1) [reply src/gui/menu_configuration.cpp:151]: Just keeping you on your toes!

**Comment by trisyoungs** [comment]
Agreed.  We'll leave this as-is for now, and work the GUI reorganisation to address the issue.

**Review by trisyoungs — APPROVED** [review]

**Pull request merged**